design/ceph/object/swift-and-keystone-integration.md

```diff
@@ -0,0 +1,293 @@
+  is enabled.
+* `[3]` This option maps directly to the `rgw_s3_auth_use_keystone` option.
+  Enabling it allows generating S3 credentials via an OpenStack API call, see the
+  [docs](https://docs.ceph.com/en/octopus/radosgw/keystone/#keystone-integration-with-the-s3-api).
+   If not given, the defaults of the corresponding RGW option apply.
+* `[*]` These options map directly to [RGW configuration
+  options](https://docs.ceph.com/en/octopus/radosgw/config-ref/#swift-settings),
+  the corresponding RGW option is formed by prefixing it with
+  `rgw_swift_` and replacing upper case letters by their lower case
+  letter followed by an underscore. E.g. `urlPrefix` maps to
+  `rgw_swift_url_prefix`. They are optional. If not given, the defaults
+  of the corresponding RGW option apply.
+
+The access to the Swift API is granted by creating a subuser of an RGW
+user. While commonly the access is granted via projects
+mapped from Keystone, explicit creation of subusers is supported by
+extending the `cephobjectstoreuser` resource with a new optional section
+`spec.subUsers`:
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephObjectStoreUser
+metadata:
+  name: my-user
+  namespace: rook-ceph
+spec:
+  store: my-store
+  displayName: my-display-name
+  quotas:
+    maxBuckets: 100
+    maxSize: 10G
+    maxObjects: 10000
+  capabilities:
+    user: "*"
+    bucket: "*"
+  subUsers:
+  - name: swift  [1]
+    access: full [2]
+```
+
+Annotations:
+* `[1]` This is the name of the subuser without the `username:` prefix
+  (see below for more explanation). The `name` must be unique within
+  the `subUsers` list.
+
+* `[2]` The possible values are: `read`, `write`, `readwrite`,
+  `full`. These values take their meanings from the possible values of
+  the `--access-level` option of `radosgw-admin subuser create`, as
+  documented in the [radosgw admin guide](
+  https://docs.ceph.com/en/octopus/radosgw/admin/#create-a-subuser).
+
+* `name` and `access` are required for each item in `subUsers`.
+
+The subusers are not mapped to a separate CR for the
+following reasons:
+
+* The full subuser names are prefixed with the username like
+  `my-user:subusername`, so being unique within the CR guarantees
+  global uniqueness.
+
+  Unlike `radosgw-admin` the subuser name in the CRD must be provided
+  *without the prefix* (radosgw-admin allows both, to omit or include
+  the prefix).
+
+* The subuser abstraction is very simple – only a name and an access
+  level can be configured, so a separate resource would not be
+  appropriate complexity wise.
+
+Like for the S3 access keys for the users, the swift keys created for the
+sub-users will be automatically injected into Secret objects. The
+credentials for the subusers are mapped to separate secrets, in the
+case of the example the following secret will be created:
+```yaml
+apiVersion:
+kind: Secret
+metadata:
+  name: rook-ceph-object-subuser-my-store-my-user:swift [1]
```

Comment:
I am not 100% sure but I guess `:` is not allowed to name k8s objects

Reply:
I checked this beforehand, and could not find such a restriction in general, the [overview docs]( (https://kubernetes.io/docs/concepts/overview/working-with-objects/names/) just state that some resource types restrict there names according to certain lexical rules, but no general rules are given. I'll check again whether I'll find more precise documentation. The [design document](https://github.com/kubernetes/design-proposals-archive/blob/main/architecture/identifiers.md) does also not specify which characters are allowed in object names. It says that the names are used in the API URLs, but a colon should be allowed in the path of a URL.

Reply:
Okay I just tested it, and it turns out you are right:
```
The Secret “foo:bar” is invalid: metadata.name: Invalid value: “foo:bar”: a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, ‘-‘ or ‘.’, and must start and end with an alphanumeric character (e.g. ‘example.com’, regex used for validation is ‘a-z0-9?(.a-z0-9?)*’)
```
It seems secrets are one of the resource types that *do* restrict the name to valid host-names. This then raises the question how to map the sub-user names to secret names injectively.

Reply:
Okay I think I won't try to make the mapping injective after all: The existing secrets for the user credentials already have a similar problem: there the name of the store and the name of the user are joined with a dash and can lead to colliding secret names for different users.

So I'll change the mapping by replacing the colon by a dash (and document the potential problem).